oximeter/db/src/client.rs

```diff
@@ -3338,4 +3680,524 @@ mod tests {
+
+        // We'll test moving from version 1, which just creates a database and
+        // table, to version 2, which adds two columns to that table in
+        // different SQL files.
+        client.execute(format!("CREATE DATABASE {test_name};")).await.unwrap();
+        client
+            .execute(format!(
+                "\
+            CREATE TABLE {test_name}.tbl (\
+                `col0` UInt8 \
+            )\
+            ENGINE = MergeTree()
+            ORDER BY `col0`;\
+        "
+            ))
+            .await
+            .unwrap();
+
+        // Write out the upgrading SQL files.
+        //
+        // Note that all of these statements are going in the version 2 schema
+        // directory.
+        let (schema_dir, version_dirs) =
+            create_test_upgrade_schema_directory(replicated, &[NEXT_VERSION])
+                .await;
+        const NEXT_VERSION: u64 = 2;
+        let first_sql =
+            format!("ALTER TABLE {test_name}.tbl ADD COLUMN `col1` UInt16;");
+        let second_sql =
+            format!("ALTER TABLE {test_name}.tbl ADD COLUMN `col2` String;");
+        let all_sql = [first_sql, second_sql];
+        let version_dir = &version_dirs[0];
+        for (i, sql) in all_sql.iter().enumerate() {
+            let path = version_dir.join(format!("up-{i}.sql"));
+            fs::write(path, sql)
+                .await
+                .expect("failed to write out upgrade SQL file");
+        }
+
+        // Apply the upgrade itself.
+        client
+            .apply_one_schema_upgrade(
+                replicated,
+                NEXT_VERSION,
+                schema_dir.path(),
+            )
+            .await
+            .expect("Failed to apply one schema upgrade");
+
+        // Check that it actually worked!
+        let body = client
+            .execute_with_body(format!(
+                "\
+            SELECT name, type FROM system.columns \
+            WHERE database = '{test_name}' AND table = 'tbl' \
+            ORDER BY name \
+            FORMAT CSV;\
+        "
+            ))
+            .await
+            .unwrap();
+        let mut lines = body.lines();
+        assert_eq!(lines.next().unwrap(), "\"col0\",\"UInt8\"");
+        assert_eq!(lines.next().unwrap(), "\"col1\",\"UInt16\"");
+        assert_eq!(lines.next().unwrap(), "\"col2\",\"String\"");
+        assert!(lines.next().is_none());
+    }
+
+    #[tokio::test]
+    async fn test_apply_one_schema_upgrade_replicated() {
+        const TEST_NAME: &str = "test_apply_one_schema_upgrade_replicated";
+        let logctx = test_setup_log(TEST_NAME);
+        let log = &logctx.log;
+        let mut cluster = ClickHouseCluster::new()
+            .await
+            .expect("Failed to initialise ClickHouse Cluster");
+        let address = cluster.replica_1.address;
+        test_apply_one_schema_upgrade_impl(log, address, true).await;
+
+        // TODO-cleanup: These should be arrays.
```

Comment:
If this TODO is meant for a follow up PR, can it be linked to the corresponding issue?

Reply:
Done in 6b3d86622